### ISSUE TITLE
Upgrade from v2022.28.1 (fresh .ova download) to v2024.4.1 Fails

### ISSUE THREAD

**Issue by jrjobe** [REMnux/remnux-cli#168]
Running remnux upgrade fails.  First tried it with the VirtualBox ova image, then downloaded the general image to try on VMware and it still fails.  This is what it ends with:

Update returned exit code not zero
Error: Update returned exit code not zero
    at ChildProcess.<anonymous> (/snapshot/remnux-cli/remnux-cli.js:571:23)
    at ChildProcess.emit (events.js:315:20)
    at maybeClose (internal/child_process.js:1021:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)

[saltstack.log](https://github.com/REMnux/remnux-cli/files/14450955/saltstack.log)

**Comment by digitalsleuth** [comment]
Hi @jrjobe , this issue is due to a recent upgrade to m2crypto which removes support for Python2. I'll be submitting a PR this evening to fix this issue, once I get back to my computer. Everything else installed correctly though.

Cheers!

**Comment by jrjobe** [comment]
Thank you and I appreciate you looking into this!

**Comment by damadorPL** [comment]
> remnux-cli@1.3.4.2.g87c65ef
> remnux-version: v2023.20.1

> downloading v2024.9.2
>> downloading remnux-salt-states-v2024.9.2.tar.gz.asc
>> downloading remnux-salt-states-v2024.9.2.tar.gz.sha256
>> downloading remnux-salt-states-v2024.9.2.tar.gz.sha256.asc
>> downloading remnux-salt-states-v2024.9.2.tar.gz
> validating file remnux-salt-states-v2024.9.2.tar.gz
> validating signature for remnux-salt-states-v2024.9.2.tar.gz.sha256
> extracting update remnux-salt-states-v2024.9.2.tar.gz
> using previous mode: dedicated
> upgrading/updating to v2024.9.2
>> Log file: /var/cache/remnux/cli/v2024.9.2/saltstack.log

Update returned exit code not zero
Error: Update returned exit code not zero
    at ChildProcess.<anonymous> (/snapshot/remnux-cli/remnux-cli.js:569:23)
    at ChildProcess.emit (events.js:315:20)
    at maybeClose (internal/child_process.js:1021:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)


Sometimes problems occur due to network or server issues when
downloading packages, in which case retrying your operation
a bit later might lead to good results.

To determine the nature of the issue, please review the
saltstack.log file under /var/cache/remnux/cli in the
subdirectory that matches the REMnux version you're installing.
Search for the log file for "result: false" messages and look at
the surrounding lines to diagnose the issue.

For assistance go to https://github.com/REMnux/remnux-cli/issues


error still persist

**Comment by damadorPL** [comment]
cat saltstack.log 
/usr/bin/salt-call:6: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  from pkg_resources import load_entry_point
[DEBUG   ] Missing configuration file: /etc/salt/minion
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: remnux
[DEBUG   ] Using importlib_metadata to load entry points
[ERROR   ] 'EntryPoints' object has no attribute 'items'
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/salt/utils/parsers.py", line 210, in parse_args
    mixin_after_parsed_func(self)
  File "/usr/lib/python3/dist-packages/salt/utils/parsers.py", line 887, in __setup_extended_logging
    log.setup_extended_logging(self.config)
  File "/usr/lib/python3/dist-packages/salt/log/setup.py", line 414, in setup_extended_logging
    providers = salt.loader.log_handlers(opts)
  File "/usr/lib/python3/dist-packages/salt/loader/__init__.py", line 686, in log_handlers
    _module_dirs(
  File "/usr/lib/python3/dist-packages/salt/loader/__init__.py", line 148, in _module_dirs
    for entry_point in entrypoints.iter_entry_points("salt.loader"):
  File "/usr/lib/python3/dist-packages/salt/utils/entrypoints.py", line 43, in _wrapped
    return f(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/utils/entrypoints.py", line 62, in iter_entry_points
    for entry_point_group, entry_points_list in entry_points.items():
AttributeError: 'EntryPoints' object has no attribute 'items'
Usage: salt-call [options] <function> [arguments]

salt-call: error: Error while processing <function LogLevelMixIn.__setup_extended_logging at 0x7f5a40946b80>: Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/salt/utils/parsers.py", line 210, in parse_args
    mixin_after_parsed_func(self)
  File "/usr/lib/python3/dist-packages/salt/utils/parsers.py", line 887, in __setup_extended_logging
    log.setup_extended_logging(self.config)
  File "/usr/lib/python3/dist-packages/salt/log/setup.py", line 414, in setup_extended_logging
    providers = salt.loader.log_handlers(opts)
  File "/usr/lib/python3/dist-packages/salt/loader/__init__.py", line 686, in log_handlers
    _module_dirs(
  File "/usr/lib/python3/dist-packages/salt/loader/__init__.py", line 148, in _module_dirs
    for entry_point in entrypoints.iter_entry_points("salt.loader"):
  File "/usr/lib/python3/dist-packages/salt/utils/entrypoints.py", line 43, in _wrapped
    return f(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/utils/entrypoints.py", line 62, in iter_entry_points
    for entry_point_group, entry_points_list in entry_points.items():
AttributeError: 'EntryPoints' object has no attribute 'items'

**Comment by digitalsleuth** [comment]
Hi @damadorPL , since you're upgrading from a version that's quite a bit older, I would first recommend downloading the latest installer from [here](https://github.com/REMnux/remnux-cli). Once you've done that, this issue can then be resolved by running the following:

`sudo python3 -m pip install "importlib_metadata<5.0.0"`

After that finishes, you should be able to run the installer again. 

Cheers!

**Comment by roelandjansen** [comment]
@digitalsleuth   --  got the same. from the fresh ova. 

I would suggest to have the ova file instead.

**Comment by digitalsleuth** [comment]
@roelandjansen You should be able to run the same command, then install again:
`sudo python3 -m pip install "importlib_metadata<5.0.0"`

Until such time as the OVA is updated, this will resolve your current issue.

Cheers

**Comment by cloud9-kj** [comment]
I've downloaded the latest installer from [here](https://github.com/REMnux/remnux-cli), and executed 
sudo python3 -m pip install "importlib_metadata<5.0.0"
Still getting the same error as above

Update returned exit code not zero
Error: Update returned exit code not zero
at ChildProcess. (/snapshot/remnux-cli/remnux-cli.js:569:23)
at ChildProcess.emit (events.js:315:20)
at maybeClose (internal/child_process.js:1021:16)
at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)

**Comment by digitalsleuth** [comment]
Hi @cloud9-kj , can  you send us the saltstack.log file from your installation attempt so we can review what may have occurred?

**Comment by cloud9-kj** [comment]
@digitalsleuth Hello, thanks for the quick response, I have attached the logs. 
[Uploading saltstack.log…]()
[Uploading saltstack_1.log…]()
[Uploading saltstack_2.log…]()

**Comment by cloud9-kj** [comment]
[Uploading saltstack.log…]()

**Comment by digitalsleuth** [comment]
Hi @cloud9-kj , unfortunately none of the links you've provided actually link to a file, but only link back to this issue. Can you try again?

**Comment by kryponix** [comment]
Hi,
  Let me attach it (as I'm having the exactly same issues and also on the last remnux-cli). Many thaks.

[saltstack.log](https://github.com/REMnux/remnux-cli/files/15066508/saltstack.log)

**Comment by digitalsleuth** [comment]
Hi @kryponix, you can try running the following commands to fix your issue, then attempt again:
```bash
sudo touch /usr/local/lib/python3.8/dist-packages/lxml-4.6.3.dist-info/METADATA
sudo touch /usr/local/lib/python3.8/dist-packages/olefile-0.46.dist-info/METADATA
sudo touch /usr/local/lib/python3.8/dist-packages/protobuf-3.20.3.dist-info/METADATA
```

This issue previously occurred, and was present in the OVA, due to an issue with the versions of pip, setuptools, and wheel which were used to install those pacakges. This will essentially create blank files named METADATA in those directories, allowing the packages to be properly uninstalled, re-installed, and upgraded.

Please let us know how it goes!

**Comment by kryponix** [comment]
Hi @digitalsleuth ,
  Thanks a lot for your quick help.
  It didn't go well :-). Same error still. I attach the new salstack log.
  [saltstack.log](https://github.com/REMnux/remnux-cli/files/15075516/saltstack.log)

  Does it mean I need to update pip, setuptools and wheel? Currently in versions:
 pip Version: 24.0
 wheel Version: 0.40.0
 setuptools Version: 67.7.2

**Comment by digitalsleuth** [comment]
You don't need to change those, however it looks like the RECORD file is also missing from those paths as well. You can rerun the same commands above, but instead replace METADATA with RECORD and try again.

**Comment by cloud9-kj** [comment]
I ran all the commands and checked RECORDS, but I'm getting the same error..

![saltstack.log.txt](https://github.com/REMnux/remnux-cli/files/15076880/saltstack.log.txt)

**Comment by cloud9-kj** [comment]
I ran all the commands and checked RECORDS, but I'm getting the same error..

![saltstack.log.txt](https://github.com/REMnux/remnux-cli/files/15076880/saltstack.log.txt)

**Comment by cloud9-kj** [comment]
![saltstack1.log.txt](https://github.com/REMnux/remnux-cli/files/15076958/saltstack1.log.txt)

**Comment by digitalsleuth** [comment]
Hi @cloud9-kj , There seems to be two errors which are occurring. The first is an issue with connectivity / DNS resolution:

`[ERROR   ] Failed to cache http://snapshots.mitmproxy.org/6.0.2/mitmproxy-6.0.2-linux.tar.gz: Error: [Errno -3] Temporary failure in name resolution reading /6.0.2/mitmproxy-6.0.2-linux.tar.gz`

This can occur sometimes if you're behind a proxy, or if the site itself is unavailable. I would recommend double-checking to see whether or not you are behind a proxy.

The second is an issue with the draios repo being outdated on your system. You can run the following and it should be resolved on the next run:

`sudo rm /etc/apt/sources.list.d/draios.list`

Please let us know how this goes!

**Comment by sscoconutree** [comment]
I'm having the same issue as well. I downloaded a fresh OVA from Remnux website. Any working solutions to fix this? I am having an error like on this post whenever I do "remnux upgrade"

**Comment by digitalsleuth** [comment]
> I'm having the same issue as well. I downloaded a fresh OVA from Remnux website. Any working solutions to fix this? I am having an error like on this post whenever I do "remnux upgrade"

Hi @sscoconutree , can you provide us with a copy of your saltstack.log file to troubleshoot? Cheers!

**Comment by kryponix** [comment]
> You don't need to change those, however it looks like the RECORD file is also missing from those paths as well. You can rerun the same commands above, but instead replace METADATA with RECORD and try again.

Thanks. In my case, after creating those empty files it worked.

**Comment by sscoconutree** [comment]
Hi @digitalsleuth here's my issue. This is a fresh OVA from the Remnux website.:

![image](https://github.com/REMnux/remnux-cli/assets/59388557/935f1c60-ad6a-4c64-8840-39529163e6d0)

I have added these prior to running `remnux upgrade`:

`sudo touch /usr/local/lib/python3.8/dist-packages/lxml-4.6.3.dist-info/METADATA`
`sudo touch /usr/local/lib/python3.8/dist-packages/olefile-0.46.dist-info/METADATA`
`sudo touch /usr/local/lib/python3.8/dist-packages/protobuf-3.20.3.dist-info/METADATA`

It seems it generated 2 saltstack logs after I ran `remnux upgrade` so I'll just provide both just in case:

Here's my saltstack log for v2022.28.1:

https://github.com/sscoconutree/saltstack_logs_remnux/raw/main/saltstack_v2022_28_1.log

Here's my saltstack log for v2024.9.2:

https://github.com/sscoconutree/saltstack_logs_remnux/raw/main/saltstack_v2024_9_2.log

Thanks!

**Comment by digitalsleuth** [comment]
Hi @sscoconutree , it looks like another file, a `RECORD` entry is missing for `olefile` and `protobuf`,  as well. You can try this:
`sudo touch /usr/local/lib/python3.8/dist-packages/olefile-0.46.dist-info/RECORD`
`sudo touch /usr/local/lib/python3.8/dist-packages/protobuf-3.20.3.dist-info/RECORD`

And that should work.
It also appears you had a temporary connectivity issue which prevented a couple of files (`extract_swf.py` and `ghidra-data-type.zip`) from being downloaded as well. These were only based on your network connection, and should be fine as well.

**Comment by Dcrash0veride** [comment]
I'm just commenting in case this helps someone else. I spent my afternoon trying to get remnux update to work and followed the advice from several threads to no avail. These steps included:

1. Import appliance into vbox
2. sudo rm /etc/apt/sources.list.d/draios.list (or whatever, tab completion ftw)
3. sudo apt-get update && sudo apt-get upgrade
4. remnux update (no work)
5. Follow download steps for new installer release (curl, chmod, mv)
6. I had the metadata, record, and other files available, so I didn't touch them
7. sudo rm /etc/apt/source.list.d/draios.list again
8. remnux update (no work)
9. remnux version (1.3.4.2?)
10. redo remnux installer
11. remnux version (1.3.9)
12. sudo apt-get update && sudo apt-get upgrade
13. remnux update (1.3.9 version used) no work
14. remnux version (1.3.4.2!?)
15. mulligan
16. sudo rm /etc/apt/sources.list,d/draios.list or whatever
17. sudo apt-get update && sudo apt-get upgrade
18. remnux upgrade completed successfully.

tl;dr I was never able to get remnux update to work, but I was able to skip it and go straight to remnux upgrade and that seems to have worked. It looks like each time I ran remnux update it replaced my remnux-cli-version and the draios thing came back. Hoping this works for now.

**Comment by roelandjansen** [comment]
So we have an upgrade flow that fails and/or requires a lot of work for multiple people.

I would suggest to update the image first so that any new user won't have this issue (and drops remnux).

Secondly, fixing the issue also would help.

Now I see this too much as "nice try, we go on"
